deeprank/features/Edesolv.py

```diff
@@ -0,0 +1,411 @@
+
+try:
+    from Bio.PDB import PDBParser, SASA
+
+except ImportError:
+    warnings.warn('Biopython module not found. Make sure you have Biopython 1.79 or later')
+    raise
+
+
+class Edesolv(FeatureClass):
+
+    # init the class
+    def __init__(self, pdb_data):
+        """Compute the energy of desolvation feature.
+
+        Biopython 1.79 or later is required for this feature.
+
+        Args:
+            pdb_data (list(byte) or str): pdb data
+
+        Example:
+            >>> edesolv = Edesolv('1AK4.pdb')
+            >>> edesolv.get_feature()
+        """
+        self.pdb = pdb_data
+        print('PDB DATA: ')
+        print(pdb_data)
+        print('TYPE: ' + str(type(pdb_data)))
+
+
+        self.feature_data = {}
+        self.feature_data_xyz = {}
+    
+    # the feature extractor
+    def get_feature(self, verbose=False):
+        """ Computes the actual Edesolv feature"""
+        
+        esolcpx = 0.0
+        esolfree = 0.0
+        
+        # create a sql database
+        pdb_db = pdb2sql(self.pdb)
+        self.db = interface(pdb_db)
+        
+        #Create a temporary pdb-like string to calculate complex and free Edesolv
+        temp_pdb = StringIO(('\n').join(pdb_db.sql2pdb()))
+        temp_pdb.seek(0)
+
+        # Parse structure in Bio.PDB
+        p = PDBParser(QUIET=1) 
+        struct = p.get_structure('temp', temp_pdb)
+        
+        # Identify chains in the Bio.PDB object
+        chains = [x.id for x in struct[0] if x.id != ' ']
+        chainA = chains[0]
+        chainB = chains[1]
+        
+        # Make free_structure fake object and translate the chains away from each other
+        # TODO: use pdb2sql.translation for this
+        free_struct = deepcopy(struct)
+        for i, chain in enumerate(chains):
+            for residue in free_struct[0][chain]:
+                for atom in residue:
+                    if i == 0:
+                        atom.coord += 300
+                    elif i == 0:
+                        atom.coord -= 300
+        
+        
+        # get the contact atoms
+        indA,indB = list(self.db.get_contact_atoms(chain1=chainA, chain2=chainB).values())
+        contact = indA + indB
+        
+        # extract the atom keys and xyz of the contact CA atoms
+        keys = self.db.get('serial,chainID,resName,resSeq,name',rowID=contact)
+        xyz = self.db.get('x,y,z',rowID=contact)
+        
+        #Make SASA class
+        sr = SASA.ShrakeRupley()
+        #Compute complex structue SASA
+        sr.compute(struct, level='A') 
+        #Compute free structure SASA
+        sr.compute(free_struct, level='A')
+
+
+        # Get disulfide bonds (important to determine sulfide atoms edesolv values)
+        disulfides_cys = get_disulfide_bonds(self.db)
+
+        
+        # create the dictionary of human readable and xyz-val data
+        self.edesolv_data = {}
+        self.edesolv_data_xyz = {}
+
+        for key, coords in zip(keys, xyz):
```

Comment:
The loop here is using the contact atoms to calculate Edsolv. The contact atoms are dependent on distance cutoff. So the Edsolv will also be dependent on distance cutoff. It is not a great idea to calculate Edsolv in this way. 
You could keep these code and rewrite one as the way used in Haddock, then compare the results to see if the Pearson correlation can very closed to 1 or not.

Reply:
So would you want me to calculate the Edesolv for each atom in the proteins (as HADDOCK does), although in the end we only use the ones in the grid? Do you want it to be done constitutively or just for the comparison with HADDOCK?

Reply:
Aha, I forgot we need only the atoms in the grid. It's fine then to calculate the Edesolv for only contact atoms with one place to improve:
Make sure the distance_cutoff for defining contact is same as the one used to get grid atoms. You should transfer the value in `DataGenerator.create_database(contact_distance=8.5)` to the "# get the contact atoms" step in this code.

To make sure the results are correct, you also need to compare the results with Haddock by giving a big enough distanct cutoff to get all atoms.

Reply:
Ok, that makes sense.
Thank you very much for the tips, will do.